tools/syz-declextract/syz-declextract.cpp

```diff
@@ -162,41 +171,163 @@ class Matcher : public MatchFinder::MatchCallback {
+    }
+    std::string filename =
+        std::filesystem::path(
+            Result.SourceManager->getFilename(decl->getCanonicalDecl()->getSourceRange().getBegin()).str())
+            .filename()
+            .stem()
+            .string();
+    std::replace(filename.begin(), filename.end(), '-', '_');
+    return std::make_optional(decl->getNameAsString() + "$auto_" + filename);
+  }
+
+  std::vector<Ops> getOps(const MatchFinder::MatchResult &Result, const std::string opsName, const InitListExpr *init) {
+    ASTContext *context = Result.Context;
+    std::vector<Ops> cmds;
+    const auto n_ops = init->getInit(genlFamilyMember["n_" + opsName])->getIntegerConstantExpr(*context);
+    const auto &opsRef = init->getInit(genlFamilyMember[opsName])->getAsBuiltinConstantDeclRef(*context);
+    if (!n_ops.has_value() || !opsRef) {
+      return {};
+    }
+    const auto *opsDecl = llvm::dyn_cast<VarDecl>(opsRef);
+    if (!opsDecl->getInit()) {
+      // NOTE: This usually happens when the ops is defined as an extern variable
+      // TODO: Extract extern variables
+      return {};
+    }
+    const auto *opsInit = llvm::dyn_cast<InitListExpr>(opsDecl->getInit());
+    for (const auto &field : opsInit->getInit(0)->getType()->getAsRecordDecl()->fields()) {
+      opsMember[opsName][field->getNameAsString()] = field->getFieldIndex();
+    }
+    for (int i = 0; i < n_ops; ++i) {
+      const auto &init = llvm::dyn_cast<InitListExpr>(opsInit->getInit(i));
+      const auto &cmdInit = init->getInit(opsMember[opsName]["cmd"])->getEnumConstantDecl();
+      if (!cmdInit) {
+        continue;
+      }
+      const auto &cmd = cmdInit->getNameAsString();
+      const ValueDecl *policyDecl = nullptr;
+      if (opsName != "small_ops") {
+        policyDecl = init->getInit(opsMember[opsName]["policy"])->getAsBuiltinConstantDeclRef(*context);
+      }
+      cmds.push_back({std::move(cmd), getPolicyName(Result, policyDecl)});
+    }
+    return cmds;
+  }
+
+  void genlFamily(const MatchFinder::MatchResult &Result) {
+    ASTContext *context = Result.Context;
+    const auto *genlFamily = Result.Nodes.getNodeAs<RecordDecl>("genl_family");
+    if (!genlFamily) {
+      return;
+    }
+    for (const auto &field : genlFamily->fields()) {
+      genlFamilyMember[field->getNameAsString()] = field->getFieldIndex();
+    }
+    const auto *genlFamilyInit = Result.Nodes.getNodeAs<InitListExpr>("genl_family_init");
+    if (!genlFamilyInit) {
+      return;
+    }
+
+    auto name = llvm::dyn_cast<StringLiteral>(genlFamilyInit->getInit(genlFamilyMember["name"]))->getString().str();
+    const auto &globalPolicyName =
+        genlFamilyInit->getInit(genlFamilyMember["policy"])->getAsBuiltinConstantDeclRef(*context);
+
+    std::string familyPolicyName;
+    if (globalPolicyName) {
+      familyPolicyName = getPolicyName(Result, globalPolicyName).value();
+    }
+
+    std::replace(name.begin(), name.end(), '.', '_'); // Illegal character.
```

Comment:
This better moved right after the name variable is initialized. This location looks rather random for this operation.

Reply:
Done!